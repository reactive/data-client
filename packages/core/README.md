<h1>
<div align="center">
<a href="https://dataclient.io" target="_blank" rel="noopener">
  <img alt="Reactive Data Client" src="./data_client_logo_and_text.svg?sanitize=true">
</a>
</div>
</h1>

[![CircleCI](https://circleci.com/gh/reactive/data-client/tree/master.svg?style=shield)](https://circleci.com/gh/reactive/data-client)
[![Coverage Status](https://img.shields.io/codecov/c/gh/reactive/data-client/master.svg?style=flat-square)](https://app.codecov.io/gh/reactive/data-client?branch=master)
[![npm downloads](https://img.shields.io/npm/dt/@data-client/core.svg?style=flat-square)](https://www.npmjs.com/package/@data-client/core)
[![bundle size](https://img.shields.io/bundlephobia/minzip/@data-client/core?style=flat-square)](https://bundlephobia.com/result?p=@data-client/core)
[![npm version](https://img.shields.io/npm/v/@data-client/core.svg?style=flat-square)](https://www.npmjs.com/package/@data-client/core)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)

Reducer/flux normalized, framework-agnostic data store. Includes managers/middleware, global referential equality guarantees,
automatic expiry policies, data normalization. Consumes [TypeScript Standard Endpoints](https://www.npmjs.com/package/@data-client/endpoint)

<div align="center">

**[📖Read The Docs](https://dataclient.io/docs)** &nbsp;|&nbsp; [🏁Getting Started](https://dataclient.io/docs/getting-started/installation) &nbsp;|&nbsp;
[🎮Todo Demo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/todo-app?file=src%2Fpages%2FHome%2FTodoList.tsx) &nbsp;|&nbsp;
[🎮Github Demo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/github-app?file=src%2Fpages%2FIssueList.tsx) &nbsp;|&nbsp;
[🎮NextJS SSR Demo](https://stackblitz.com/github/reactive/data-client/tree/master/examples/nextjs?file=components%2Ftodo%2FTodoList.tsx)

</div>

### Framework Implementations

- [React](https://www.npmjs.com/package/@data-client/react)
- [React-Redux](https://dataclient.io/docs/guides/redux)

### Sample React Hook suspense implementation

```typescript
function useSuspense(endpoint, ...args)
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { data, expiryStatus, expiresAt } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  const error = controller.getError(endpoint, ...args, state);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    return controller.fetch(endpoint, ...args);
    // we need to check against serialized params, since params can change frequently
  }, [expiresAt, controller, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend even if it is not fresh
  if (expiryStatus !== ExpiryStatus.Valid && maybePromise) {
    throw maybePromise;
  }

  if (error) throw error;

  return data;
}
```


## API

- [Controller](https://dataclient.io/docs/api/Controller)
  - [ctrl.fetch](https://dataclient.io/docs/api/Controller#fetch)
  - [ctrl.fetchIfStale](https://dataclient.io/docs/api/Controller#fetchIfStale)
  - [ctrl.expireAll](https://dataclient.io/docs/api/Controller#expireAll)
  - [ctrl.invalidate](https://dataclient.io/docs/api/Controller#invalidate)
  - [ctrl.invalidateAll](https://dataclient.io/docs/api/Controller#invalidateAll)
  - [ctrl.resetEntireStore](https://dataclient.io/docs/api/Controller#resetEntireStore)
  - [ctrl.set](https://dataclient.io/docs/api/Controller#set)
  - [ctrl.setResponse](https://dataclient.io/docs/api/Controller#setResponse)
  - [ctrl.setError](https://dataclient.io/docs/api/Controller#setError)
  - [ctrl.resolve](https://dataclient.io/docs/api/Controller#resolve)
  - [ctrl.subscribe](https://dataclient.io/docs/api/Controller#subscribe)
  - [ctrl.unsubscribe](https://dataclient.io/docs/api/Controller#unsubscribe)
  - [ctrl.get](https://dataclient.io/docs/api/Controller#get)
  - [ctrl.getResponse](https://dataclient.io/docs/api/Controller#getResponse)
  - [ctrl.getError](https://dataclient.io/docs/api/Controller#getError)
  - [ctrl.snapshot](https://dataclient.io/docs/api/Controller#snapshot)
  - [ctrl.getState](https://dataclient.io/docs/api/Controller#getState)
- Middleware: [LogoutManager](https://dataclient.io/docs/api/LogoutManager), [NetworkManager](https://dataclient.io/docs/api/NetworkManager), [SubscriptionManager](https://dataclient.io/docs/api/SubscriptionManager), [PollingSubscription](https://dataclient.io/docs/api/PollingSubscription), [DevToolsManager](https://dataclient.io/docs/api/DevToolsManager)
- State: createReducer(), initialState, applyManager