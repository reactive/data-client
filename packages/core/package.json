{
  "name": "@data-client/core",
  "version": "0.14.18",
  "description": "Async State Management without the Management. REST, GraphQL, SSE, Websockets, Fetch",
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "next": [
        "lib/next/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "next": [
        "ts3.4/next/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./next": {
      "types": "./lib/next/index.d.ts",
      "require": "./dist/next.js",
      "default": "./lib/next/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md",
    "./typescript.svg",
    "./data_client_logo_and_text.svg"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' yarn g:babel --out-dir lib",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' yarn g:babel --out-dir legacy",
    "build:js:node": "BROWSERSLIST_ENV=node12 yarn g:rollup",
    "build:js:browser": "BROWSERSLIST_ENV=legacy yarn g:rollup",
    "build:bundle": "yarn g:runs build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "yarn g:clean ./index.d.ts",
    "build:legacy-types": "yarn g:downtypes lib ts3.4",
    "build": "run build:lib && run build:legacy:lib && run build:bundle",
    "dev": "run build:lib -w",
    "prepare": "run build:lib",
    "prepack": "run prepare && run build:bundle && run build:legacy:lib"
  },
  "keywords": [
    "data",
    "cache",
    "normalizr",
    "normalization",
    "flux",
    "fetch",
    "hook",
    "networking",
    "async",
    "typescript",
    "async",
    "data fetching",
    "data cache",
    "reactive",
    "state management",
    "api client",
    "api",
    "normalized cache",
    "swr",
    "query",
    "ios",
    "android",
    "web",
    "middleware",
    "websocket",
    "SSE",
    "GraphQL",
    "REST",
    "RPC",
    "declarative"
  ],
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "funding": "https://github.com/sponsors/ntucker",
  "license": "Apache-2.0",
  "homepage": "https://dataclient.io",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:reactive/data-client.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/reactive/data-client/issues"
  },
  "dependencies": {
    "@babel/runtime-corejs3": "^7.26.0",
    "@data-client/normalizr": "^0.14.16",
    "core-js-pure": "^3.40.0",
    "flux-standard-action": "^2.1.1"
  },
  "devDependencies": {
    "@anansi/browserslist-config": "^1.4.2",
    "@data-client/endpoint": "workspace:*",
    "@types/jest": "^29.5.14",
    "@types/node": "^22.0.0",
    "rollup-plugins": "workspace:*"
  }
}
