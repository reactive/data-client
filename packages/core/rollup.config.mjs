import {
  babel,
  commonjs,
  filesize,
  json,
  resolve,
  replace,
  terser,
  typeConfig,
  typeConfigNext,
  onwarn,
} from 'rollup-plugins';

import pkg from './package.json' with { type: 'json' };

const dependencies = Object.keys(pkg.dependencies)
  //.concat(Object.keys(pkg.peerDependencies))
  .filter(dep => !['@babel/runtime'].includes(dep));

const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node', '.jsx'];
const nativeExtensions = ['.native.ts', ...extensions];
process.env.NODE_ENV = 'production';

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];
if (process.env.BROWSERSLIST_ENV !== 'node12') {
  // browser-friendly UMD build
  configs.push({
    input: 'src/index.ts',
    output: [{ file: pkg.unpkg, format: 'umd', name: 'RDC.Core' }],
    onwarn,
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        rootMode: 'upward',
        babelHelpers: 'runtime',
        caller: { polyfillMethod: false },
      }),
      replace({
        'process.env.NODE_ENV': JSON.stringify('production'),
        preventAssignment: true,
      }),
      resolve({ extensions }),
      commonjs({ extensions }),
      json(),
      terser({}),
      filesize({ showBrotliSize: true }),
    ],
  });
  configs.push(typeConfig);
  configs.push(typeConfigNext);
} else {
  // node-friendly commonjs build
  [
    { input: 'src/index.ts', output: pkg.main },
    { input: 'src/next/index.ts', output: 'dist/next.js' },
  ].forEach(({ input, output }) => {
    configs.push({
      input,
      external: isExternal,
      output: [{ file: output, format: 'cjs' }],
      onwarn,
      plugins: [
        babel({
          exclude: ['node_modules/**', '**/__tests__/**', '**/*.d.ts'],
          extensions: nativeExtensions,
          rootMode: 'upward',
          babelHelpers: 'runtime',
        }),
        replace({ 'process.env.CJS': 'true', preventAssignment: true }),
        resolve({ extensions: nativeExtensions }),
        commonjs({ extensions: nativeExtensions }),
      ],
    });
  });
}
export default configs;
