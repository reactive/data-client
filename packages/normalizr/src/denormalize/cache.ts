import type { EntityInterface } from '../interface.js';
import { EntityPath } from '../types.js';
import type { INVALID } from './symbol.js';

export default interface Cache {
  getEntity(
    pk: string,
    schema: EntityInterface,
    entity: any,
    computeValue: (localCacheKey: Map<string, any>) => void,
  ): object | undefined | typeof INVALID;
  getResults(
    input: any,
    cachable: boolean,
    computeValue: () => any,
  ): {
    data: any;
    paths: EntityPath[];
  };
}
