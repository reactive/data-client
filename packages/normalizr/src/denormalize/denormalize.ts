import { getEntities } from './getEntities.js';
import LocalCache from './localCache.js';
import getUnvisit from './unvisit.js';
import type { Schema } from '../interface.js';
import { isImmutable } from '../schemas/ImmutableUtils.js';
import type { DenormalizeNullable } from '../types.js';
import type { INVALID } from './symbol.js';

export function denormalize<S extends Schema>(
  schema: S | undefined,
  input: any,
  entities: any,
  args: readonly any[] = [],
): DenormalizeNullable<S> | typeof INVALID {
  // undefined means don't do anything
  if (schema === undefined || input === undefined) {
    return input as any;
  }

  return getUnvisit(
    getEntities(entities),
    new LocalCache(),
    args,
    isImmutable(entities),
  )(schema, input).data;
}
