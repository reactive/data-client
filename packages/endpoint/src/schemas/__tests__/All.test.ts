// eslint-env jest
import { initialState, State, Controller } from '@data-client/core';
import {
  normalize,
  MemoCache,
  denormalize as plainDenormalize,
  INVALID,
  MemoPolicy as PojoDelegate,
} from '@data-client/normalizr';
import {
  MemoPolicy as ImmDelegate,
  denormalize as immDenormalize,
} from '@data-client/normalizr/imm';
import { IDEntity } from '__tests__/new';

import { schema } from '../..';
import { fromJSState } from './denormalize';

let dateSpy: jest.SpyInstance<number, []>;
beforeAll(() => {
  dateSpy = jest

    .spyOn(global.Date, 'now')
    .mockImplementation(() => new Date('2019-05-14T11:01:58.135Z').valueOf());
});
afterAll(() => {
  dateSpy.mockRestore();
});

describe.each([[]])(`${schema.All.name} normalization (%s)`, () => {
  test('should throw a custom error if data loads with string unexpected value', () => {
    class User extends IDEntity {}
    const sch = new schema.All(User);
    function normalizeBad() {
      normalize(sch, 'abc');
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('should throw a custom error if data loads with json string unexpected value', () => {
    class User extends IDEntity {}
    const sch = new schema.All(User);
    function normalizeBad() {
      normalize(sch, '[{"id":5}]');
    }
    expect(normalizeBad).toThrowErrorMatchingSnapshot();
  });

  test('normalizes Objects using their values', () => {
    class User extends IDEntity {}
    const { result, entities } = normalize(new schema.All(User), {
      foo: { id: '1' },
      bar: { id: '2' },
    });
    expect(result).toBeUndefined();
    expect(entities).toMatchSnapshot();
  });

  describe('Class', () => {
    class Cats extends IDEntity {}
    test('normalizes a single entity', () => {
      const listSchema = new schema.All(Cats);
      expect(
        normalize(listSchema, [{ id: '1' }, { id: '2' }]).entities,
      ).toMatchSnapshot();
    });

    test('normalizes multiple entities', () => {
      const inferSchemaFn = jest.fn(input => input.type || 'dogs');
      class Person extends IDEntity {}
      const listSchema = new schema.All(
        {
          Cat: Cats,
          people: Person,
        },
        inferSchemaFn,
      );

      const { result, entities } = normalize(listSchema, [
        { type: 'Cat', id: '123' },
        { type: 'people', id: '123' },
        { id: '789', name: 'fido' },
        { type: 'Cat', id: '456' },
      ]);
      expect(result).toBeUndefined();
      expect(entities).toMatchSnapshot();
      expect(inferSchemaFn.mock.calls).toMatchSnapshot();
    });

    test('normalizes Objects using their values', () => {
      class User extends IDEntity {}
      const users = new schema.All(User);
      expect(
        normalize(users, { foo: { id: '1' }, bar: { id: '2' } }).entities,
      ).toMatchSnapshot();
    });

    test('filters out undefined and null normalized values', () => {
      class User extends IDEntity {}
      const users = new schema.All(User);
      expect(
        normalize(users, [undefined, { id: '123' }, null]).entities,
      ).toMatchSnapshot();
    });
  });
});

describe.each([
  [
    'direct',
    <T>(data: T) => data,
    <T>(data: T) => data,
    PojoDelegate,
    plainDenormalize,
  ],
  [
    'immutable',
    fromJSState,
    (v: any) => (typeof v?.toJS === 'function' ? v.toJS() : v),
    ImmDelegate,
    immDenormalize,
  ],
])(
  `${schema.Array.name} denormalization (%s)`,
  (_, createInput, createOutput, MyDelegate, denormalize) => {
    test('denormalizes a single entity', () => {
      class Cat extends IDEntity {}
      const state: State<unknown> = createInput({
        ...initialState,
        entities: {
          Cat: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        },
        indexes: {},
      }) as any;
      const sch = new schema.All(Cat);
      expect(
        new Controller({ memo: new MemoCache(MyDelegate) }).get(sch, state),
      ).toMatchSnapshot();
    });

    test('denormalizes nested in object', () => {
      class Cat extends IDEntity {}
      const catSchema = { results: new schema.All(Cat) };
      const state = createInput({
        entities: {
          Cat: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        },
        indexes: {},
      });
      // use memocache because we don't support 'object' schemas in controller yet
      expect(
        new MemoCache(MyDelegate).query(catSchema, [], state).data,
      ).toMatchSnapshot();
    });

    test('denormalizes nested in object with primitive', () => {
      class Cat extends IDEntity {}
      const catSchema = { results: new schema.All(Cat), nextPage: '' };
      const state = createInput({
        entities: {
          Cat: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        },
        indexes: {},
      });
      const value = new MemoCache(MyDelegate).query(catSchema, [], state).data;
      expect(value).not.toEqual(expect.any(Symbol));
      if (typeof value === 'symbol' || value === undefined) return;
      expect(createOutput(value.results)).toMatchSnapshot();
      expect(createOutput(value)).toMatchSnapshot();
    });

    test('denormalizes removes undefined or INVALID entities', () => {
      class Cat extends IDEntity {}
      const catSchema = { results: new schema.All(Cat), nextPage: '' };
      const state = createInput({
        entities: {
          Cat: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
            3: undefined,
            4: INVALID,
          },
        },
        indexes: {},
      });
      const value = new MemoCache(MyDelegate).query(catSchema, [], state).data;
      expect(value).not.toEqual(expect.any(Symbol));
      if (typeof value === 'symbol' || value === undefined) return;
      expect(createOutput(value.results).length).toBe(2);
      expect(createOutput(value.results)).toMatchSnapshot();
      expect(createOutput(value)).toMatchSnapshot();
    });

    test('denormalize maintains referential equality until entities are added', () => {
      class Cat extends IDEntity {}
      (Cat as any).defaults;
      const catSchema = { results: new schema.All(Cat), nextPage: '' };
      let state: State<unknown> = {
        ...initialState,
        entities: {
          Cat: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        },
        indexes: {},
      };
      const memo = new MemoCache();
      const value = memo.query(catSchema, [], state).data;

      expect(createOutput(value).results?.length).toBe(2);
      expect(createOutput(value).results).toMatchSnapshot();
      const value2 = memo.query(catSchema, [], state).data;
      expect(createOutput(value).results[0]).toBe(
        createOutput(value2).results[0],
      );
      expect(value).toBe(value2);

      state = {
        ...state,
        entities: {
          ...state.entities,
          Cat: {
            ...state.entities.Cat,
            3: { id: '3', name: 'Jelico' },
          },
        },
      };
      const value3 = memo.query(catSchema, [], state).data;
      expect(createOutput(value3).results?.length).toBe(3);
      expect(createOutput(value3).results).toMatchSnapshot();
      expect(createOutput(value).results[0]).toBe(
        createOutput(value3).results[0],
      );
      expect(createOutput(value).results[2]).not.toBe(
        createOutput(value3).results[2],
      );
      expect(value).not.toBe(value3);
    });

    test('denormalizes should be invalid when no entities are present', () => {
      class Cat extends IDEntity {}
      const catSchema = { results: new schema.All(Cat) };
      const state = createInput({
        entities: {
          DOG: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        },
        indexes: {},
      });
      const value = new MemoCache(MyDelegate).query(catSchema, [], state).data;
      expect(createOutput(value)).toEqual(expect.any(Symbol));
    });

    test('denormalizes should not be found when no entities are present (polymorphic)', () => {
      class Cat extends IDEntity {
        readonly type = 'Cat';
      }
      class Dog extends IDEntity {
        readonly type = 'dogs';
      }
      class Person extends IDEntity {
        readonly type = 'people';
      }
      const listSchema = new schema.All(
        {
          Cat: Cat,
          dogs: Dog,
          people: Person,
        },
        input => input.type || 'dogs',
      );

      const state = createInput({
        entities: {
          DOG: {
            1: { id: '1', name: 'Milo' },
            2: { id: '2', name: 'Jake' },
          },
        },
        indexes: {},
      });
      const value = new MemoCache(MyDelegate).query(listSchema, [], state).data;
      expect(createOutput(value)).toEqual(expect.any(Symbol));
    });

    test('returns the input value if is null', () => {
      class Filling extends IDEntity {}
      class Taco extends IDEntity {
        static schema = { fillings: new schema.All(Filling) };
      }
      const state = createInput({
        ...initialState,
        entities: {
          Taco: {
            123: {
              id: '123',
              fillings: null,
            },
          },
        },
      });
      expect(denormalize(Taco, '123', state.entities)).toMatchSnapshot();
    });

    test('denormalizes multiple entities', () => {
      class Cat extends IDEntity {
        readonly type = 'Cat';
      }
      class Dog extends IDEntity {
        readonly type = 'dogs';
      }
      class Person extends IDEntity {
        readonly type = 'people';
      }
      const listSchema = new schema.All(
        {
          Cat: Cat,
          dogs: Dog,
          people: Person,
        },
        input => input.type || 'dogs',
      );

      const state = createInput({
        entities: {
          Cat: {
            123: {
              id: '123',
              type: 'Cat',
            },
            456: {
              id: '456',
              type: 'Cat',
            },
          },
          Person: {
            123: {
              id: '123',
              type: 'people',
            },
          },
        },
        indexes: {},
      });
      const value = new MemoCache(MyDelegate).query(listSchema, [], state).data;
      expect(value).not.toEqual(expect.any(Symbol));
      if (typeof value === 'symbol') return;
      expect(value).toMatchSnapshot();
      const first = value && value[0];
      // type check to ensure correct inference
      first?.type;
    });

    test('does not allow initializing with non-entities', () => {
      class Cat extends IDEntity {}
      const catRecord = new schema.Object({
        cat: Cat,
      });
      // @ts-expect-error
      const catList = new schema.All(catRecord);
    });
  },
);
