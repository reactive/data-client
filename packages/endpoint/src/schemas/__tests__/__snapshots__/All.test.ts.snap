// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AllSchema normalization (%s) Class filters out undefined and null normalized values 1`] = `
{
  "User": {
    "123": {
      "id": "123",
    },
  },
}
`;

exports[`AllSchema normalization (%s) Class normalizes Objects using their values 1`] = `
{
  "User": {
    "1": {
      "id": "1",
    },
    "2": {
      "id": "2",
    },
  },
}
`;

exports[`AllSchema normalization (%s) Class normalizes a single entity 1`] = `
{
  "Cats": {
    "1": {
      "id": "1",
    },
    "2": {
      "id": "2",
    },
  },
}
`;

exports[`AllSchema normalization (%s) Class normalizes multiple entities 1`] = `
{
  "Cats": {
    "123": {
      "id": "123",
      "type": "Cat",
    },
    "456": {
      "id": "456",
      "type": "Cat",
    },
  },
  "Person": {
    "123": {
      "id": "123",
      "type": "people",
    },
  },
}
`;

exports[`AllSchema normalization (%s) Class normalizes multiple entities 2`] = `
[
  [
    {
      "id": "123",
      "type": "Cat",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "123",
      "type": "Cat",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "123",
      "type": "people",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "123",
      "type": "people",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "789",
      "name": "fido",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "789",
      "name": "fido",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "456",
      "type": "Cat",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
  [
    {
      "id": "456",
      "type": "Cat",
    },
    [
      {
        "id": "123",
        "type": "Cat",
      },
      {
        "id": "123",
        "type": "people",
      },
      {
        "id": "789",
        "name": "fido",
      },
      {
        "id": "456",
        "type": "Cat",
      },
    ],
    undefined,
  ],
]
`;

exports[`AllSchema normalization (%s) Class normalizes multiple entities 3`] = `
[
  [
    "Schema attribute "dogs" is not expected.
Expected one of: "Cat", "people"

Value: {
  "id": "789",
  "name": "fido"
}",
  ],
]
`;

exports[`AllSchema normalization (%s) normalizes Objects using their values 1`] = `
{
  "User": {
    "1": {
      "id": "1",
    },
    "2": {
      "id": "2",
    },
  },
}
`;

exports[`AllSchema normalization (%s) should throw a custom error if data loads with json string unexpected value 1`] = `
"Normalizing a string, but this does match schema.

Parsing this input string as JSON worked. This likely indicates fetch function did not parse
the JSON. By default, this only happens if "content-type" header includes "json".
See https://dataclient.io/rest/api/RestEndpoint#parseResponse for more information

  Schema: [
  {
    "key": "User",
    "schema": {}
  }
]
  Input: "[{"id":5}]""
`;

exports[`AllSchema normalization (%s) should throw a custom error if data loads with string unexpected value 1`] = `
"Unexpected input given to normalize. Expected type to be "object", found "string".

          Schema: [
  {
    "key": "User",
    "schema": {}
  }
]
          Input: "abc""
`;

exports[`ArraySchema denormalization (direct) denormalize maintains referential equality until entities are added 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (direct) denormalize maintains referential equality until entities are added 2`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
  Cat {
    "id": "3",
    "name": "Jelico",
  },
]
`;

exports[`ArraySchema denormalization (direct) denormalizes a single entity 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (direct) denormalizes multiple entities 1`] = `
[
  Cat {
    "id": "123",
    "type": "Cat",
  },
  Cat {
    "id": "456",
    "type": "Cat",
  },
  Person {
    "id": "123",
    "type": "people",
  },
]
`;

exports[`ArraySchema denormalization (direct) denormalizes nested in object 1`] = `
{
  "results": [
    Cat {
      "id": "1",
      "name": "Milo",
    },
    Cat {
      "id": "2",
      "name": "Jake",
    },
  ],
}
`;

exports[`ArraySchema denormalization (direct) denormalizes nested in object with primitive 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (direct) denormalizes nested in object with primitive 2`] = `
{
  "nextPage": "",
  "results": [
    Cat {
      "id": "1",
      "name": "Milo",
    },
    Cat {
      "id": "2",
      "name": "Jake",
    },
  ],
}
`;

exports[`ArraySchema denormalization (direct) denormalizes removes undefined or INVALID entities 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (direct) denormalizes removes undefined or INVALID entities 2`] = `
{
  "nextPage": "",
  "results": [
    Cat {
      "id": "1",
      "name": "Milo",
    },
    Cat {
      "id": "2",
      "name": "Jake",
    },
  ],
}
`;

exports[`ArraySchema denormalization (direct) denormalizes removes undefined or INVALID entities with polymorphism 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
    "type": "Cat",
  },
  Cat {
    "id": "2",
    "name": "Jake",
    "type": "Cat",
  },
  Dog {
    "id": "1",
    "name": "Rex",
    "type": "dogs",
  },
  Person {
    "id": "1",
    "name": "Alice",
    "type": "people",
  },
]
`;

exports[`ArraySchema denormalization (direct) returns the input value if is null 1`] = `
Taco {
  "fillings": null,
  "id": "123",
}
`;

exports[`ArraySchema denormalization (immutable) denormalize maintains referential equality until entities are added 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (immutable) denormalize maintains referential equality until entities are added 2`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
  Cat {
    "id": "3",
    "name": "Jelico",
  },
]
`;

exports[`ArraySchema denormalization (immutable) denormalizes a single entity 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (immutable) denormalizes multiple entities 1`] = `
[
  Cat {
    "id": "123",
    "type": "Cat",
  },
  Cat {
    "id": "456",
    "type": "Cat",
  },
  Person {
    "id": "123",
    "type": "people",
  },
]
`;

exports[`ArraySchema denormalization (immutable) denormalizes nested in object 1`] = `
{
  "results": [
    Cat {
      "id": "1",
      "name": "Milo",
    },
    Cat {
      "id": "2",
      "name": "Jake",
    },
  ],
}
`;

exports[`ArraySchema denormalization (immutable) denormalizes nested in object with primitive 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (immutable) denormalizes nested in object with primitive 2`] = `
{
  "nextPage": "",
  "results": [
    Cat {
      "id": "1",
      "name": "Milo",
    },
    Cat {
      "id": "2",
      "name": "Jake",
    },
  ],
}
`;

exports[`ArraySchema denormalization (immutable) denormalizes removes undefined or INVALID entities 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
  },
  Cat {
    "id": "2",
    "name": "Jake",
  },
]
`;

exports[`ArraySchema denormalization (immutable) denormalizes removes undefined or INVALID entities 2`] = `
{
  "nextPage": "",
  "results": [
    Cat {
      "id": "1",
      "name": "Milo",
    },
    Cat {
      "id": "2",
      "name": "Jake",
    },
  ],
}
`;

exports[`ArraySchema denormalization (immutable) denormalizes removes undefined or INVALID entities with polymorphism 1`] = `
[
  Cat {
    "id": "1",
    "name": "Milo",
    "type": "Cat",
  },
  Cat {
    "id": "2",
    "name": "Jake",
    "type": "Cat",
  },
  Dog {
    "id": "1",
    "name": "Rex",
    "type": "dogs",
  },
  Person {
    "id": "1",
    "name": "Alice",
    "type": "people",
  },
]
`;

exports[`ArraySchema denormalization (immutable) returns the input value if is null 1`] = `
Taco {
  "fillings": null,
  "id": "123",
}
`;
